// angular-dynamic-loader/esm5/lib/dynamic-loader.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
var DynamicLoaderService = /** @class */ (function () {
    function DynamicLoaderService() {
        this.subject = new BehaviorSubject(null);
    }
    DynamicLoaderService.prototype.getLoaderData = function () {
        return this.subject.asObservable();
    };
    DynamicLoaderService.prototype.show = function () {
        this.subject.next({ show: true });
    };
    DynamicLoaderService.prototype.hide = function () {
        this.subject.next({ show: false });
    };
    DynamicLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DynamicLoaderService_Factory() { return new DynamicLoaderService(); }, token: DynamicLoaderService, providedIn: "root" });
    DynamicLoaderService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], DynamicLoaderService);
    return DynamicLoaderService;
}());
export { DynamicLoaderService };
