// @ckeditor/ckeditor5-angular/esm2015/ckeditor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */
/**
 * Basic typings for the CKEditor5 elements.
 */
export var CKEditor5;
(function (CKEditor5) {
    /**
     * The CKEditor5 editor constructor.
     * @record
     */
    function EditorConstructor() { }
    CKEditor5.EditorConstructor = EditorConstructor;
    if (false) {
        /**
         * @param {?} sourceElementOrData
         * @param {?=} config
         * @return {?}
         */
        EditorConstructor.prototype.create = function (sourceElementOrData, config) { };
    }
    /**
     * The CKEditor5 editor config.
     * @record
     */
    function Config() { }
    CKEditor5.Config = Config;
    /**
     * The event object passed to CKEditor5 event callbacks.
     *
     * See https://ckeditor.com/docs/ckeditor5/latest/api/module_utils_eventinfo-EventInfo.html
     * @record
     * @template EventName
     */
    function EventInfo() { }
    CKEditor5.EventInfo = EventInfo;
    if (false) {
        /** @type {?} */
        EventInfo.prototype.name;
        /** @type {?} */
        EventInfo.prototype.path;
        /** @type {?} */
        EventInfo.prototype.source;
        /** @type {?|undefined} */
        EventInfo.prototype.return;
        /**
         * @return {?}
         */
        EventInfo.prototype.off = function () { };
        /**
         * @return {?}
         */
        EventInfo.prototype.stop = function () { };
    }
    /**
     * The base Editor class.
     *
     * See https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editor-Editor.html
     * @record
     */
    function BaseEditor() { }
    CKEditor5.BaseEditor = BaseEditor;
    if (false) {
        /** @type {?} */
        BaseEditor.prototype.commands;
        /** @type {?} */
        BaseEditor.prototype.config;
        /** @type {?} */
        BaseEditor.prototype.conversion;
        /** @type {?} */
        BaseEditor.prototype.data;
        /** @type {?} */
        BaseEditor.prototype.editing;
        /** @type {?} */
        BaseEditor.prototype.keystrokes;
        /** @type {?} */
        BaseEditor.prototype.locale;
        /** @type {?} */
        BaseEditor.prototype.model;
        /** @type {?} */
        BaseEditor.prototype.plugins;
        /** @type {?} */
        BaseEditor.prototype.state;
        /** @type {?} */
        BaseEditor.prototype.isReadOnly;
        /* Skipping unhandled member: [ property: string ]: any;*/
        /**
         * @param {?} events
         * @return {?}
         */
        BaseEditor.prototype.delegate = function (events) { };
        /**
         * @return {?}
         */
        BaseEditor.prototype.destroy = function () { };
        /**
         * @param {?} commandName
         * @param {...?} params
         * @return {?}
         */
        BaseEditor.prototype.execute = function (commandName, params) { };
        /**
         * @param {?} eventName
         * @param {?=} args
         * @return {?}
         */
        BaseEditor.prototype.fire = function (eventName, args) { };
        /**
         * @param {?} emitter
         * @param {?} eventName
         * @param {?} callback
         * @param {?=} options
         * @return {?}
         */
        BaseEditor.prototype.listenTo = function (emitter, eventName, callback, options) { };
        /**
         * @param {?} eventName
         * @param {?} callback
         * @return {?}
         */
        BaseEditor.prototype.off = function (eventName, callback) { };
        /**
         * @param {?} eventName
         * @param {?} callback
         * @param {?=} options
         * @return {?}
         */
        BaseEditor.prototype.on = function (eventName, callback, options) { };
        /**
         * @param {?} eventName
         * @param {?} callback
         * @param {?=} options
         * @return {?}
         */
        BaseEditor.prototype.once = function (eventName, callback, options) { };
        /**
         * @param {?} emitter
         * @param {?} eventName
         * @param {?} callback
         * @return {?}
         */
        BaseEditor.prototype.stopListening = function (emitter, eventName, callback) { };
        /**
         * @param {...?} args
         * @return {?}
         */
        BaseEditor.prototype.t = function (args) { };
    }
    /**
     * The CKEditor5 DataApi interface.
     *
     * See https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_utils_dataapimixin-DataApi.html.
     * @record
     */
    function DataApi() { }
    CKEditor5.DataApi = DataApi;
    if (false) {
        /**
         * @return {?}
         */
        DataApi.prototype.getData = function () { };
        /**
         * @param {?} data
         * @return {?}
         */
        DataApi.prototype.setData = function (data) { };
    }
    /**
     * A CKEditor5 editor that implements the
     * [DataApi interface](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_utils_dataapimixin-DataApi.html).
     * E.g. the `ClassicEditor`, `InlineEditor`, etc.
     * @record
     */
    function Editor() { }
    CKEditor5.Editor = Editor;
    /**
     * @record
     */
    function ContextWatchdog() { }
    CKEditor5.ContextWatchdog = ContextWatchdog;
    if (false) {
        /** @type {?} */
        ContextWatchdog.prototype.context;
        /** @type {?} */
        ContextWatchdog.prototype._watchdogs;
        /**
         * @param {?} items
         * @return {?}
         */
        ContextWatchdog.prototype.add = function (items) { };
        /**
         * @param {?} items
         * @return {?}
         */
        ContextWatchdog.prototype.remove = function (items) { };
        /**
         * @param {?} itemId
         * @return {?}
         */
        ContextWatchdog.prototype.getItem = function (itemId) { };
        /**
         * @param {?} itemId
         * @param {?} itemType
         * @param {?} watchdog
         * @return {?}
         */
        ContextWatchdog.prototype.addItemWatchdog = function (itemId, itemType, watchdog) { };
    }
    /**
     * @record
     */
    function EditorWatchdog() { }
    CKEditor5.EditorWatchdog = EditorWatchdog;
    if (false) {
        /** @type {?} */
        EditorWatchdog.prototype.editor;
    }
    /**
     * @record
     * @template T
     */
    function Watchdog() { }
    CKEditor5.Watchdog = Watchdog;
    if (false) {
        /**
         * @param {?} creator
         * @return {?}
         */
        Watchdog.prototype.setCreator = function (creator) { };
        /**
         * @param {?} destructor
         * @return {?}
         */
        Watchdog.prototype.setDestructor = function (destructor) { };
        /**
         * @param {?} event
         * @param {?} callback
         * @return {?}
         */
        Watchdog.prototype.on = function (event, callback) { };
        /**
         * @return {?}
         */
        Watchdog.prototype.destroy = function () { };
        /**
         * @param {...?} args
         * @return {?}
         */
        Watchdog.prototype.create = function (args) { };
    }
})(CKEditor5 || (CKEditor5 = {}));
